plugins/teststeps/teststeps.go

```diff
@@ -29,94 +30,66 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	type tgtErr struct {
-		target *target.Target
-		err    error
+	// Temporary solution
+	ctx, cancelCtx := combineChannelsToContext(cancel, pause)
+	defer cancelCtx()
+
+	reportTarget := func(t *target.Target, err error) {
+		if err != nil {
+			log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, t, err)
+			select {
+			case ch.Err <- cerrors.TargetError{Target: t, Err: err}:
+			case <-ctx.Done():
+				log.Debugf("%s: ForEachTarget: received cancellation/pause signal while reporting error", pluginName)
+			}
+		} else {
+			log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, t)
+			select {
+			case ch.Out <- t:
+			case <-ctx.Done():
```

Comment:
Can this race if we are cancelling? (Not sure it is a problem if it can race, just asking)

Reply:
I think not, at least the behaviour is the same as before (previously we had

select {
case ch.Out <- t:
case <-pause:
case <-cancel:
}